src/contracts/orderbook/NiftyswapOrderbook.sol

```diff
@@ -0,0 +1,414 @@
+    }
+
+    /**
+     * Purchases a token.
+     * @param listingId The ID of the listing.
+     * @param quantity The quantity of tokens to purchase.
+     * @param additionalFees The additional fees to pay.
+     * @param additionalFeeRecievers The addresses to send the additional fees to.
+     * @dev Royalties are taken from the listing cost.
+     */
+    function acceptListing(
+        bytes32 listingId,
+        uint256 quantity,
+        uint256[] memory additionalFees,
+        address[] memory additionalFeeRecievers
+    ) external {
+        Order memory listing = orders[listingId];
+        if (!listing.isListing || listing.creator == address(0)) {
+            // Is a listing, cancelled, completed or never existed
+            revert InvalidListingId(listingId);
+        }
+
+        _acceptOrder(listingId, listing, quantity, additionalFees, additionalFeeRecievers);
+
+        emit ListingAccepted(listingId, msg.sender, listing.tokenContract, quantity);
+    }
+
+    /**
+     * Sells a token.
+     * @param offerId The ID of the listing.
+     * @param quantity The quantity of tokens to sell.
+     * @param additionalFees The additional fees to pay.
+     * @param additionalFeeRecievers The addresses to send the additional fees to.
+     */
+    function acceptOffer(
+        bytes32 offerId,
+        uint256 quantity,
+        uint256[] memory additionalFees,
+        address[] memory additionalFeeRecievers
+    ) external {
+        Order memory offer = orders[offerId];
+        if (offer.isListing || offer.creator == address(0)) {
+            // Is a listing, cancelled, completed or never existed
+            revert InvalidOfferId(offerId);
+        }
+
+        _acceptOrder(offerId, offer, quantity, additionalFees, additionalFeeRecievers);
+
+        emit OfferAccepted(offerId, msg.sender, offer.tokenContract, quantity);
+    }
+
+    function _acceptOrder(
+        bytes32 orderId,
+        Order memory order,
+        uint256 quantity,
+        uint256[] memory additionalFees,
+        address[] memory additionalFeeRecievers
+    ) internal {
+        if (quantity == 0 || quantity > order.quantity) {
+            revert InvalidQuantity();
+        }
+        if (_isExpired(order)) {
+            revert InvalidExpiry();
+        }
+        if (additionalFees.length != additionalFeeRecievers.length) {
+            revert InvalidAdditionalFees();
+        }
+
+        // Calculate payables
+        uint256 totalCost = order.pricePerToken * quantity;
+        (address royaltyRecipient, uint256 royaltyAmount) =
+            getRoyaltyInfo(order.tokenContract, order.tokenId, totalCost);
+
+        address currencyReceiver = order.isListing ? order.creator : msg.sender;
+        address tokenReceiver = order.isListing ? msg.sender : order.creator;
+
+        if (royaltyAmount > 0) {
+            // Transfer royalties
+            TransferHelper.safeTransferFrom(order.currency, tokenReceiver, royaltyRecipient, royaltyAmount);
```

Comment:
Maybe the order should have maximum `royalty` to protect the user from an malicious token that asks for a royalty that drains all balance?

Reply:
@PhABC has asked to leave it uncapped in case the rate is high for a legitimate reason (e.g. apple 30% royalty)

Reply:
Oh I don't mean a hard cap on the whole contract, but instead a cap per order. When you create the order you explicitly say how much you are willing to pay on royalties.

Keep in mind this is just a `safeTransferFrom` meaning that a malicious nft could even drain the account of the user.

Reply:
Oh nvm, `remainingCost` underflows and the tx reverts.